### PR TITLE
Get typed system values

### PR DESCRIPTION
**Pull request by nickvergessen**
Methods `getSystemValue(Bool|Int|String)()` have been added.
They use the empty value of their type, have the return type aswell as the default type typehinted.

Then this is used for the maintenance mode as an example, because there was a problem if you had `'maintenance' => 'false',` in your config. Even `occ maintenance:mode --off` failed to correctly turn it off.

I could also imagine a type map being added for our known configs, which would throw an error, if you further on get a typed config via the generic method: `getSystemValue('typed-config')`. In the same run, we could also use this to define a list of default values for all those values, so an app dev can not use a wrong default behaviour causing weird behaviours later on: `getSystemValueBool('maintenance', true)`

Ref #3301

### CODE REVIEW

File: lib/private/AllConfig.php [hunk 1]
```diff
@@ -125,6 +125,42 @@ public function getSystemValue($key, $default = '') {
...
+	 * @param mixed $default the default value to be returned if the value isn't set
+	 * @return mixed the value or $default
+	 * @since 16.0.0
+	 */
+	public function getSystemValueBool(string $key, bool $default = false): bool {
+		return (bool) $this->getSystemValue($key, $default);
+	}
+
+	/**
+	 * Looks up a integer system wide defined value
```
Comment (nickvergessen) [inline lib/private/AllConfig.php]:
```suggestion
	 * Looks up an integer system wide defined value
```

File: ocs/v1.php [hunk 2]
```diff
@@ -33,7 +33,7 @@
 require_once __DIR__ . '/../lib/base.php';
 
 if (\OCP\Util::needUpgrade()
-	|| \OC::$server->getSystemConfig()->getValue('maintenance', false)) {
+	|| \OC::$server->getConfig()->getSystemValueBool('maintenance')) {
```
Comment (MorrisJobke) [inline ocs/v1.php]: I think this was made on purpose to have no need to setup the DB connection and stuff like that in the error case.

Comment (nickvergessen) [reply ocs/v1.php]: Yes, but it also doesn't set connection, if you use getSystemValue* only.
It's not in the constructor:
https://github.com/nextcloud/server/blob/a55af5191203764378964ac3fd7664af3e5163c9/lib/private/AllConfig.php#L73-L76


https://github.com/nextcloud/server/blob/a55af5191203764378964ac3fd7664af3e5163c9/lib/private/AllConfig.php#L91-L95

fixDIInit is only called for *User* and *App* will set up the database too, so I guess it's fine?

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍 Fine with me

**Review by rullzer — APPROVED** [review]

**Pull request merged**